R/DIFFUSE.R

```diff
@@ -0,0 +1,266 @@
+        check4 <- .checkArgs(param, list())
+
+        if (length(c(check1,check2,check3,check4)) > 0 ) {
+            stop("Error: incorrect argument(s) for algorithm, silent, noLog, and/or param.
+                 See documentation for DIFFUSE function")
+        }
+
+        #Extract AGG metadata for writing to log
+        metaAGG <- attr(AGG,"meta")
+           #Assign AGG vertices and edges to data frames
+           AGGverts<- igraph::as_data_frame(AGG, "vertices")
+           AGGedges<- igraph::as_data_frame(AGG, "edges")
+
+           sizeEdges <- nrow(AGGedges) #Number of edges, used for indexing purposes
+           sizeVerts <- nrow(AGGverts) #Number of vertices, used for indexing purposes
+
+    if (algorithm == "Leis") {
+
+        consoleVect<- "Method Selected: Leis"
+        if (!silent) {
+            print(consoleVect)
+        }
+
+        #check beta
+        if ( length(.checkArgs(param[[1]], numeric(length = 1))) > 1) {
+            stop("Leis Method selected. Beta must be class numeric of type
+                 double")
+        }
+
+        #creating transition matrix W
+        zeroVect <- numeric(length(sizeVerts)^2) #Make a vector of zeros from which to make W
+        W<-matrix(data = zeroVect, nrow = sizeVerts , ncol = sizeVerts,
+                  dimnames = list(AGGverts$name, AGGverts$name))
+
+        vertDegree <- numeric(length = sizeVerts) #Will store degree of each vertex
+        names(vertDegree) <- AGGverts$name
+
+        consoleVect<-"Calculating degree for vertices"
+        if (!silent) {
+            print(consoleVect)
+        }
+        #calculating degree via following procedure
+
+        HGNCsymb<-AGGverts$name #get a vector of AGG vertex names
+
+        for (i in 1:sizeVerts) {
+
+            consoleVect<-paste("vertex",i, "of", sizeVerts, sep = " ")
+            if (!silent) {
+                print(consoleVect)
+            }
+
+
+            edgeIndex <- as.logical(grepl(HGNCsymb[i],AGGedges$from, fixed = TRUE)
+                                    | grepl(HGNCsymb[i],AGGedges$to, fixed = TRUE))
+            #generate a list of genes present in edges including vertex of interest
+            geneList <- c(AGGedges$from[edgeIndex],
+                             AGGedges$to[edgeIndex])
+            #reduce to unique elements
+            geneList <- unique(geneList)
+            #remove vertex of interest from list
+            geneList <- geneList[-grep(HGNCsymb[i],geneList, fixed = TRUE)]
+
+            deg <- length(geneList) #degree of vertex of interest
+
+            vertDegree[grep(HGNCsymb[i],names(vertDegree), fixed = TRUE)] <- deg
+
+        }
+
+    #Assigning values to matrix W: See Leiseron et. al 2015 for procedure description
+
+        consoleVect <-  "Constructing matrix W"
+        if (!silent) {
+            print(consoleVect)
+        }
+
+        doneEdges <- 0
+
+        for (n in 1:sizeVerts) {
+
+
+
+           jElement <- AGGverts$name[n]
```

Comment:
This is really inefficient. You can work from start to finish with indices - you never actually need to use names and thus you don't need to use grep. And since the indices are the same as the row-indices, this entire loop reduces to a single expression.